comm.go

```diff
@@ -121,6 +123,22 @@ func (p *PubSub) handleNewPeer(ctx context.Context, pid peer.ID, outgoing <-chan
 	}
 }
 
+func (p *PubSub) handleNewPeerWithBackoff(ctx context.Context, pid peer.ID, outgoing <-chan *RPC) error {
+	delay, valid := p.deadPeerBackoff.updateAndGet(pid)
+	if !valid {
+		return fmt.Errorf("backoff attempts to %s expired after reaching maximum allowed", pid)
+	}
```

Comment:
let's return the error directly from `updateAndGet` instead of a bool, makes for simpler code.

Reply:
I decoupled updating the backoff history from checking for the number of backoff attempts. Hence now `updateAndGet` [only increments](https://github.com/yhassanzadeh13/go-libp2p-pubsub/blob/7f815f0a9c4e957c775ce3c1f10a704221671284/backoff.go#L76) the backoff attempts without returning any boolean or error indicating the maximum attempts reached. Instead, I introduced a separate [`peerExceededBackoffThreshold`](https://github.com/yhassanzadeh13/go-libp2p-pubsub/blob/7f815f0a9c4e957c775ce3c1f10a704221671284/backoff.go#L82-L90) method that checks whether the backoff attempts of a peer exceeds the defined threshold. The reasons for this change are: 

1. We want to close the channel and forget the peer if the backoff attempts go beyond a threshold. The [`updateAndGet` is called on a goroutine](https://github.com/yhassanzadeh13/go-libp2p-pubsub/blob/6ebc292245fe234c3c078ef87362697ab2c1ed6a/pubsub.go#L694) while the `messages` channel and `peers` map are residing on a separate `goroutine`. So, if we let `updateAndGet` return a backoff attempt exceeding error and the `goroutine` that it resides on attempts on closing the `messages` channel and forgetting peer from `peers` map, it results in a race condition between the two `goroutines`, and also the code is vulnerable to panic when this [`defer`](https://github.com/yhassanzadeh13/go-libp2p-pubsub/blob/6ebc292245fe234c3c078ef87362697ab2c1ed6a/pubsub.go#L518-L522) function is executed, as it is trying to close an already closed channel, i.e.,the `messages` channel that we closed on `updateAndGet`. 

2. By this decoupling, we [invoke `peerExceededBackoffThreshold` on the parent `goroutine`](https://github.com/yhassanzadeh13/go-libp2p-pubsub/blob/6ebc292245fe234c3c078ef87362697ab2c1ed6a/pubsub.go#L687), hence we give up on backing off if the peer exceeds the backoff threshold rightaway, without opening any channel, spawning any child `goroutine` for `updateAndGet`. Moreover, the peer [is forgotten](https://github.com/yhassanzadeh13/go-libp2p-pubsub/blob/6ebc292245fe234c3c078ef87362697ab2c1ed6a/pubsub.go#L699) by the subsequent lines. Hence, no vulnerability for race conditions and less resource allocation-deallocations. 

Please let me know how does it sound?

Reply:
sounds better than what we had :)